Familiar AI Fix/Revamp v3

GriffonSpade opened the pull request:
Associated Thread: [Familiar AI Fix/Revamp](https://www.servuo.com/threads/familiar-ai-fix-revamp.6961/)
This pull is designed to clear out and streamline the kludge that is Familiar AI.

Before being merged to master, this pull should definitely be tested, preferably have confirmation in OSI behavior fidelity, and any necessary refinement to correct if either of the previous two have issues.

-Fixes bug that usually cause familiar to be unable to work properly for the Dark Tides quest. (Though the AI rarely takes a bit to get around the wall)
-Familiars ONLY attack their master's current combatant.
-Familiars ONLY attack creatures within RangeHome (default 10) of master.
--Sanity Check: If master is attacking familiar, or vise-versa, set attacker's combatant to null. (Masters don't attack their familiars, familiars don't attack their masters)
-No support for Archer/Mage familiars. (all familiars are Melee)
-While not attacking, familiars follow their master.
-Familiars teleport to their master if they're on another map.
-Familiars teleport nearer to their master if they're more than RangePerception (default 16) apart.

v2:
-Brought code into line with Master.
-Fixed a a bug causing familiar to not attack anything not attacking itself
--Reduced setting of OrderType to only when necessary
-Changed AggressiveAction() familiar behavior to abort early on
-Various incidental Trims

v3:
-Brought code into line with Master.
-Fixed Familiar Powersliding issues.
-Fixed AttacksMastersTarget default to false in BaseFamiliar

kevin-10 commented:
Upon aggroing, familiars are auto-attacking the target.  They should not do this except death adder and vampire bat.

GriffonSpade commented:
I believe that was not my error, but the one who did that edit: AttacksMastersTarget was set to true in BaseFamiliar (I did not change it). I thought it was odd, but I didn't want to change any of the existing functionality.

GriffonSpade commented:
Reverted and changed some of the previous stuff. I think I might have this power-sliding and warping issue buggered. At least for familiars.

As a final question, will the familiars on OSI without AttackMastersTarget attack the enemy if they're the master's target AND happen to be in melee range? (essentially just not chasing them)

I want to go over version 3 a few more times, and I'll put it up in a new PR tomorrow.

GriffonSpade commented:
Updated this to version 3.
